src/pkcs15init/pkcs15-lib.c

```diff
@@ -1329,13 +1330,16 @@ sc_pkcs15init_init_skdf(struct sc_pkcs15_card *p15card, struct sc_profile *profi
 	key_info->key_reference = 0;
 	switch (keyargs->algorithm) {
 	case SC_ALGORITHM_DES:
-		key_info->key_type = CKM_DES_ECB;
+		key_info->key_type = CKK_DES;
```

Comment:
Why is this changed? This value goes directly to PKCS#15 CardInfo.supportedAlgorithms. The specification I have says it should be chosen from and interpreted as **mechanism numbers in PKCS #11**. Please revert this change or give reference to standard where this changed.

Reply:
I changed it because it seemed a bit strange that the mechanism is set to ECB here when the keys can be used with CKM_AES_CBC mechanism as well. I though it would be better to use the generic CKK definitions to define the key type instead of the mechanism. I have taken care of this also in function myeid_fixup_supported_algos in commit f722646. There might have been some other places as well, where it seemed cleaner to use key type defintions rather than mechanisms. 

Do you think it's ok with f722646 or would it be better to go back to CKM_AES_ECB defs everywhere? Am I breaking up something else with this change?

Reply:
While inside OpenSC this change would work. The point is that this value gets written to token, and PKCS15 defines which type to use. So this breaks interop and standard compliance.

The structure is supported to announce which algorithms (mechanisms) can be used with the key. If we want to do things better, we should list all mechanisms supported with the key, and fill-in cross-reference to each of these supportedAlgorithms entries.

For simplicity and other implementation reasons I added only the ECB mechanism. But if improvement is needed, there should way to add all mechanisms the card supports and that can be used with the key in question.

Reply:
Please can you show me where this value actually gets written to token, because I don't find it? I understand that we have to use the CKM_ values in tokenInfo as defined in PKCS#15, 
and my intention has been to do so.

My idea is that it would be better to use the CKK_AES and CKK_DES3 definitons for key type internally, because when we add functionality to do crypto operations 
with the keys, there are many situations where we just want to check whether the key is AES or DES3, and it looks wrong to compare it to a mechanism type. I would then separately list the mechanisms associated with the key type into tokenInfo->supported_algos, using the CKM_ constants.

But isn't this just what we are already doing in myeid_fixup_supported_algos?

case SC_PKCS15_TYPE_SKEY_GENERIC:
	switch (skey_info->key_type | (skey_info->value_len << 16)) {
		case CKK_AES | (128 << 16):
			_add_supported_algo(profile, p15card, object, SC_PKCS15_ALGO_OP_DECIPHER|SC_PKCS15_ALGO_OP_ENCIPHER, CKM_AES_ECB, &id_aes128_ecb);
			_add_supported_algo(profile, p15card, object, SC_PKCS15_ALGO_OP_DECIPHER|SC_PKCS15_ALGO_OP_ENCIPHER, CKM_AES_CBC, &id_aes128_cbc);
			break;
		case CKK_AES | (256 << 16):
			_add_supported_algo(profile, p15card, object, SC_PKCS15_ALGO_OP_DECIPHER|SC_PKCS15_ALGO_OP_ENCIPHER, CKM_AES_ECB, &id_aes256_ecb);
			_add_supported_algo(profile, p15card, object, SC_PKCS15_ALGO_OP_DECIPHER|SC_PKCS15_ALGO_OP_ENCIPHER, CKM_AES_CBC, &id_aes256_cbc);
			break;
		}
		break;

We do use the CKM definitions here and they go into tokenInfo in sc_pkcs15_encode_tokeninfo(). This is actually your code. Shall we let each driver handle this in their own way, because cards might not support all mechanisms that are associated with specific key type? Or should we have some more generic way that would be used by default?

I verified by testing that this also works as intended. The algoritmInfo.algorithm values written to tokenInfo are the PKCS#11 mechanisms (4225 and 4226), not what we are setting in skey_info.key_type

I would prefer to not go back, but keep using the CKK_ values for key_type internally, and ensure that they are correctly "converted" to a list of CKM_ values when writing tokenInfo (and other way when reading tokenInfo). Do you agree with this? 

In skey_info we already have unsigned int algo_refs[SC_MAX_SUPPORTED_ALGORITHMS]. The value in this array reference the AlgorithmInfo entries in tokenInfo. Isn't this the cross-reference you are talking about?

Reply:
Sorry. I remembered that part incorrectly. You are right.

My first impression would have been then to just suggest putting some non-pkcs11 value there. But further digging shows that pkcs11/framework-pkcs15.c also stashes CKK_* values in the same field (though, it's supported only for non-session objects for the time being). So yeah, it actually is good to do what you are doing.

Sorry for the noise. If you ever get to rebasing things, please add the above comment in the commit message.

Reply:
No problem and thank you for checking this.

I would like to add AlgorthmInfos for AES algorithm into tokenInfo already when the card is initialized. SO-PIN is required to update tokenInfo and this is not very convenient when calling C_WrapKey and C_UnwrapKey in PKCS#11. This probably should be a card specific choice, handled in pkcs15-init overrides. But this is out of the context of this PR.